ocaml/typing/jkind.ml

```diff
@@ -818,24 +964,38 @@ end
 module Desc = struct
   type t =
```

Comment:
This doesn't seem quite right to me: this looks like the `Desc` for product kinds. I think this may want a different structure.

Reply:
I meant to write something about this, sorry.  I agree this is weird, but submit that it was already wrong and this PR makes it no worse.  The comment about modes is plainly false, to start, and vars are layouts but consts are kind, so the model/layout mismatch is already here.  This should be cleaned up, and I vaguely though that was in the scope of the ongoing refractors of this module.

Reply:
I see what you mean here, but I maintain that the old version was OK. That is, the var case (and its comment) was meant to indicate that the var case can be treated as a kind, but with all of its mode stuff set to `max`. This wasn't captured in the old datatype (perhaps foolishly), but any consumer of it can easily apply that knowledge.

This new world, though, resists this kind of interpretation. So I see the new one as worse than the old one.

Reply:
The reason I say the comment is false is that we interpret all kinds with layout sort var as the Var case here - that is, it is used to represent some vars about which we would not say "all modes will be [max]".  I think, therefore, either the old type was wrong and should have been a record containing modes and a (const layout or var), or the old type was fine but the new one is too (because whether or not we're accurately recording kinds should be remembered at the granularity of individual product components).

Reply:
Previously, we never really used the `Var` output from `get` (the only way in which `Desc.t` interacts with the outside world). In most cases, `Var` is just a fatal error. The exception is in printing, at which point we don't care about the modes.

That's still true in the new patch, actually.

So maybe here's a middle ground: keep `Desc.t` as it is, but with an invariant that a `Product` contains at least one `Var` or `Product`. (This will require an easy implementation change to guarantee the invariant.) Then consumers have reliable mode stuff for `Const` but not for the other cases -- but those other cases are error cases in everything but the printer, so it's all OK. If that doesn't work out in the future (it looks just fine for this patch), we can think about a deeper restructuring later.

Reply:
I took a [stab](https://github.com/ccasin/flambda-backend/commit/7ea66be9536ddb730f8bf1e170ec0af876e529f7) at this, but it made printing worse - something that used to be printed as `any_non_null & any_non_null` was printed as `any & any` (with `alpha` on, so these are different).  So I reverted it.

I think we'll agree that we don't actually want to print either of those, but rather `(any & any) mod non_null`.  Which sent me down a little printing rabbit hole - I think we'll probably want to do a bigger rethink of how kinds are printed to accommodate products, and have realized I don't understand some of the current choices, so I'm going to make a ticket for now and suggest we revisit in the near future.